tracing-subscriber/src/layer/layered.rs

```diff
@@ -0,0 +1,536 @@
+use tracing_core::{
+    metadata::Metadata,
+    span,
+    subscriber::{Interest, Subscriber},
+    Event, LevelFilter,
+};
+
+#[cfg(feature = "registry")]
+use crate::registry::Registry;
+use crate::{
+    filter::{self, FilterId},
+    layer::{Context, Layer},
+    registry::LookupSpan,
+};
+use std::{
+    any::{type_name, TypeId},
+    fmt,
+    marker::PhantomData,
+};
+
+/// A [`Subscriber`] composed of a `Subscriber` wrapped by one or more
+/// [`Layer`]s.
+///
+/// [`Layer`]: crate::Layer
+/// [`Subscriber`]: https://docs.rs/tracing-core/latest/tracing_core/trait.Subscriber.html
+#[derive(Clone)]
+pub struct Layered<L, I, S = I> {
+    /// The layer.
+    layer: L,
+
+    /// The inner value that `self.layer` was layered onto.
+    ///
+    /// If this is also a `Layer`, then this `Layered` will implement `Layer`.
+    /// If this is a `Subscriber`, then this `Layered` will implement
+    /// `Subscriber` instead.
+    inner: I,
+
+    // These booleans are used to determine how to combine `Interest`s and max
+    // level hints when per-layer filters are in use.
+    /// Is `self.inner` a `Registry`?
+    ///
+    /// If so, when combining `Interest`s, we want to "bubble up" its
+    /// `Interest`.
+    inner_is_registry: bool,
+
+    /// Does `self.layer` have per-layer filters?
+    ///
+    /// This will be true if:
+    /// - `self.inner` is a `Filtered`.
+    /// - `self.inner` is a tree of `Layered`s where _all_ arms of those
+    ///   `Layered`s have per-layer filters.
+    ///
+    /// Otherwise, if it's a `Layered` with one per-layer filter in one branch,
+    /// but a non-per-layer-filtered layer in the other branch, this will be
+    /// _false_, because the `Layered` is already handling the combining of
+    /// per-layer filter `Interest`s and max level hints with its non-filtered
+    /// `Layer`.
+    has_layer_filter: bool,
+
+    /// Does `self.inner` have per-layer filters?
+    ///
+    /// This is determined according to the same rules as
+    /// `has_layer_filter` above.
+    inner_has_layer_filter: bool,
```

Comment:
i'm not sure what that would look like. These are internal implementation details that are specific to the `Layered` type, and they're not going to be part of the public API. Are you proposing replacing them with three separate `enum` types like
```rust
enum InnerIsRegistry {
    Yes,
    No,
}

enum HasLayerFilter {
    Yes,
    No,
}

enum InnerHasLayerFilter {
    Yes,
    No,
}
```
because that seems kind of unnecessary to me.

i think "boolean blindness" is mainly a concern when a _function_ takes multiple boolean arguments. like, if we were going to have a public constructor like
```rust
impl<A, B, S> Layered<A, B, S> {
    pub fn new(a: A, b: B, inner_is_registry: bool, has_layer_filter: bool, inner_has_layer_filter: bool) -> Self {
        // ...
    }
}
```
i would definitely agree that passing a bunch of unnamed bools is scary since callsites would look like
```rust
Layered::new(..., false, true, false)
```
or similar. but, that's never going to happen; these are determined by the `Layered` internally when it's constructed. Because they're just _fields_, rather than being passed to a function as a big wad of bools, they _are_ always given names...so I don't really understand what we would gain from making them into enums.

Reply:
hmm, yeah. I (naively?) assumed that those enums would be better, but if you don't agree, then feel free to ignore this suggestion.